// chain-sdk/dist/util.js
'use strict';

var x509SubjectAttributes = {
  C: { array: true },
  O: { array: true },
  OU: { array: true },
  L: { array: true },
  ST: { array: true },
  STREET: { array: true },
  POSTALCODE: { array: true },
  SERIALNUMBER: { array: false },
  CN: { array: false }
};

var sanitizeX509GuardData = function sanitizeX509GuardData(guardData) {
  var keys = Object.keys(guardData);
  if (keys.length !== 1 || keys[0].toLowerCase() !== 'subject') {
    throw new Error('X509 guard data must contain exactly one key, "subject"');
  }

  var newSubject = {};
  var oldSubject = Object.values(guardData)[0];
  for (var k in oldSubject) {
    var attrib = x509SubjectAttributes[k.toUpperCase()];
    if (!attrib) {
      throw new Error('X509 guard data contains invalid subject attribute: ' + k);
    }

    var v = oldSubject[k];
    if (!attrib.array && Array.isArray(v)) {
      throw new Error('X509 guard data contains invalid array for attribute ' + k + ': ' + v.toString());
    } else if (attrib.array && !Array.isArray(v)) {
      newSubject[k] = [v];
    } else {
      newSubject[k] = v;
    }
  }

  return { subject: newSubject };
};

module.exports = {
  sanitizeX509GuardData: sanitizeX509GuardData
};
